packages/adapter-drizzle/src/lib/mysql.ts

```diff
@@ -270,21 +302,92 @@ export function MySqlDrizzleAdapter(
           )
         )
     },
+    async createAuthenticator(data: AdapterAuthenticator) {
+      await client
+        .insert(authenticatorsTable)
+        .values({ ...data, id: crypto.randomUUID() })
+
+      return await client
+        .select()
+        .from(authenticatorsTable)
+        .where(eq(authenticatorsTable.credentialID, data.credentialID))
+        .then((res) => fromDBAuthenticator(res[0]) ?? null)
+    },
+    async getAuthenticator(credentialID: string) {
+      const authenticator = await client
+        .select()
+        .from(authenticatorsTable)
+        .where(eq(authenticatorsTable.credentialID, credentialID))
+        .then((res) => (res.length ? fromDBAuthenticator(res[0]) : null))
+      return authenticator ? authenticator : null
+    },
+    async listAuthenticatorsByUserId(userId: string) {
+      return await client
+        .select()
+        .from(authenticatorsTable)
+        .where(eq(authenticatorsTable.userId, userId))
+        .then((res) => res.map(fromDBAuthenticator))
+    },
+    async updateAuthenticatorCounter(credentialID: string, newCounter: number) {
+      await client
+        .update(authenticatorsTable)
+        .set({ counter: newCounter })
+        .where(eq(authenticatorsTable.credentialID, credentialID))
+
+      return await client
+        .select()
+        .from(authenticatorsTable)
+        .where(eq(authenticatorsTable.credentialID, credentialID))
+        .then((res) => fromDBAuthenticator(res[0]) ?? null)
+    },
+  }
+}
+
+type BaseAuthenticator = InferInsertModel<
+  ReturnType<typeof defineTables>["authenticatorsTable"]
+>
+type DrizzleAuthenticator = BaseAuthenticator &
+  Required<
+    Pick<
+      BaseAuthenticator,
+      | "userId"
+      | "providerAccountId"
+      | "counter"
+      | "credentialBackedUp"
+      | "credentialID"
+      | "credentialPublicKey"
+      | "credentialDeviceType"
+    >
+  >
+
+function fromDBAuthenticator(
+  authenticator: DrizzleAuthenticator
+): AdapterAuthenticator {
+  const { transports, id, ...other } = authenticator
+
+  return {
+    ...other,
+    transports: transports || undefined,
```

Comment:
Is there a reason not to allow this to be null? Why do we need to turn null -> undefined in a bunch of places?

Imo it would be better to change this in core lib to handle both null and undefined rather than having this `fromDBAuthenticator` helper

Reply:
the `@auth/prisma-adapter` had a similar helper function to transform the object upon retrieval, which i based this drizzle pr on.
i tried putting in a helper function in `utils.ts` but don't know from where i would import `defineTables` from to infer the shape of the drizzle table.

Reply:
my point is that this could be removed altogether if the core just accepted `string | undefined | null`, instead of just `string | undefined`. We could just return the response from the db call directly. Same is true for the Prisma adapter

Reply:
Yeah its really not doing anything else, right. 

How about this? It's just loosening up the type so it shouldn't be breaking anything 

https://github.com/nextauthjs/next-auth/pull/10861

Reply:
Okay so I also removed that fn from the prisma adapter in that PR ^^. Let me know what yall think

Reply:
@jipsterk We also dropped the `aurhwnticator.id` as the primary key and made the PK out of `credentialID` and `userId` (see PR above)